lib/datasource/crate/index.ts

```diff
@@ -1,88 +1,248 @@
+
+  /// raw URL of the registry, as specified in cargo config
+  rawUrl?: string;
+
+  /// parsed URL of the registry
+  url?: URL;
+
+  /// path where the registry is cloned
+  clonePath?: string;
+}
+
+export function getIndexSuffix(lookupName: string): string[] {
   const len = lookupName.length;
 
   if (len === 1) {
-    return '1/' + lookupName;
+    return ['1', lookupName];
   }
   if (len === 2) {
-    return '2/' + lookupName;
+    return ['2', lookupName];
   }
   if (len === 3) {
-    return '3/' + lookupName[0] + '/' + lookupName;
+    return ['3', lookupName[0], lookupName];
   }
 
-  return (
-    lookupName.slice(0, 2) + '/' + lookupName.slice(2, 4) + '/' + lookupName
-  );
+  return [lookupName.slice(0, 2), lookupName.slice(2, 4), lookupName];
 }
 
 interface CrateRecord {
   vers: string;
   yanked: boolean;
 }
 
+export async function fetchCrateRecordsPayload(
+  info: RegistryInfo,
+  lookupName: string
+): Promise<string> {
+  if (info.clonePath) {
+    const path = join(info.clonePath, ...getIndexSuffix(lookupName));
+    return fs.readFile(path, { encoding: 'utf8' });
+  }
+
+  if (info.flavor === RegistryFlavor.CratesIo) {
+    const crateUrl = CRATES_IO_BASE_URL + getIndexSuffix(lookupName).join('/');
+    try {
+      return (await http.get(crateUrl)).body;
+    } catch (err) {
+      if (
+        err.statusCode === 429 ||
+        (err.statusCode >= 500 && err.statusCode < 600)
+      ) {
+        throw new ExternalHostError(err);
+      }
+    }
+  }
+
+  throw new Error(`unsupported crate registry flavor: ${info.flavor}`);
+}
+
+/// Computes the dependency URL for a crate, given
+/// registry information
+function getDependencyUrl(info: RegistryInfo, lookupName: string): string {
+  switch (info.flavor) {
+    case RegistryFlavor.CratesIo:
+      return `https://crates.io/crates/${lookupName}`;
+    case RegistryFlavor.Cloudsmith: {
+      // input: https://dl.cloudsmith.io/basic/$org/$repo/cargo/index.git
+      const tokens = info.url.pathname.split('/');
+      const org = tokens[2];
+      const repo = tokens[3];
+      return `https://cloudsmith.io/~${org}/repos/${repo}/packages/detail/cargo/${lookupName}`;
+    }
+    default:
+      return `${info.rawUrl}/${lookupName}`;
+  }
+}
+
+/// Given a Git URL, computes a semi-human-readable name for a folder in which to
+/// clone the repository.
+function cacheDirFromUrl(url: URL): string {
+  const proto = url.protocol.replace(/:$/, '');
+  const host = url.hostname;
+  const hash = hasha(url.pathname, {
+    algorithm: 'sha256',
+  }).substr(0, 7);
+
+  return `crate-registry-${proto}-${host}-${hash}`;
+}
+
+/// Fetches information about a registry, by url.
+/// If no url is given, assumes crates.io.
+/// If an url is given, assumes it's a valid Git repository
+/// url and clones it to cache.
+async function fetchRegistryInfo(
+  registryUrl: string
+): Promise<RegistryInfo | null> {
+  let url: URL;
+  try {
+    url = new URL(registryUrl);
```

Comment:
I think global URL object is deprecated. Please import from url module

Reply:
I can do it if you want to, but I don't see it being deprecated? https://nodejs.org/api/url.html#url_the_whatwg_url_api

The `require('url').Url` one definitely seems deprecated though: https://nodejs.org/api/url.html#url_legacy_urlobject

Reply:
Ok, looks like my bad. Leave it as is.